src/util.c

```diff
@@ -500,4 +500,34 @@ RRStatus parseHashSlots(char * slots, char * string)
 exit:
     RedisModule_Free(string);
     return ret;
-}
\ No newline at end of file
+}
+
+ClientState * GetClientState(RedisRaftCtx *rr, RedisModuleCtx * ctx)
+{
+    unsigned long long client_id = RedisModule_GetClientId(ctx);
+    return RedisModule_DictGetC(rr->client_state, &client_id, sizeof(client_id), NULL);
+}
+
+uint64_t ClientStateCount(RedisRaftCtx *rr)
+{
+    return RedisModule_DictSize(rr->client_state);
+}
+
+void AllocClientState(RedisRaftCtx *rr, unsigned long long client_id)
+{
+    ClientState *clientState = RedisModule_Calloc(sizeof(ClientState), 1);
+    RedisModule_DictSetC(rr->client_state, &client_id, sizeof(client_id), clientState);
+}
+
+void FreeClientState(RedisRaftCtx *rr, unsigned long long client_id)
+{
+    ClientState *state = NULL;
+
+    if (RedisModule_DictDelC(rr->client_state, &client_id,
```

Comment:
I think if we don't have a client struct here, we are in trouble (it shouldn't be possible?). Maybe, we should just assert here. We can also use assert when we add a client to dict, just to verify there is no existing client.

Reply:
sure, its validating redis, which is fine.